fetchPythonRequirements: init (fixed output pypi fetcher)

DavHau opened the pull request:
fetchPythonRequirements is a multi package fetcher for python/pypi similar to the one used by buildGoModule or fetchCrate for rust.

It downlaods python packages specified by a list of pip-style python requirements.
It requires a maximum date 'maxDate' being specified.

The result will be as if `pip download` would have been executed at the point in time specified by maxDate.

This is ensured by putting pip behind a local proxy filtering the api responses from pypi.org to only contain files for which the release date is lower than the specified maxDate.

In other words, the proxy provides a snapshot-like view on pypi.org

The fetcher is reproducible without the need of specifying package versions or hashes. Any pip compatible requirement specifiers are allowed.

Here an example on how this fetcher can be used to build jupyterlab.
This environment, containing more than 40 python packages, can be described in less than 30 lines of nix code:

```nix
let
  pkgs = import <nixpkgs> {};
  python = pkgs.python38;
  downloaded = python.pkgs.fetchPythonRequirements {
    requirements = [
      "jupyterlab"
    ];
    maxDate = "2020-03";
    outputHash = "08h3a606kfijmm73h90dybi9pkldc4421ffxzvp47vw05ijrr2rj";
  };
  multiplePackages = python.pkgs.buildPythonPackage {
    name = "bunch-of-python-pkgs";
    format = "";
    src = downloaded;
    buildInputs = pkgs.pythonManylinuxPackages.manylinux1;
    nativeBuildInputs = [ pkgs.autoPatchelfHook python.pkgs.wheelUnpackHook ];
    unpackPhase = "mkdir dist && cp $src/* dist/";
    installPhase = ''
      mkdir -p "$out/${python.sitePackages}"
      export PYTHONPATH="$out/${python.sitePackages}:$PYTHONPATH"
      ${python}/bin/python -m pip install ./dist/*.{whl,tar.gz,zip} --no-index --no-warn-script-location --prefix="$out" --no-cache $pipInstallFlags --ignore-installed
    '';
  };
in

python.withPackages (ps: [ multiplePackages ])

```

DavHau committed:
fetchPythonRequirements: init

jonringer commented on pkgs/development/interpreters/python/fetch-python-requirements.nix:
```diff
@@ -0,0 +1,127 @@
+
+        nativeBuildInputs = with python3.pkgs; [ pythonWithMitmproxy pythonPip curl cacert ];
+
+        phases = [ "buildPhase" ];
```
IIRC, explicitly setting phases is discouraged. But don't remember the context

cc @jtojnar

jtojnar replied:
See also https://github.com/NixOS/nixpkgs/issues/28910. Mainly the issue is that overriding phases drops `fixupPhase` which breaks most setup hooks. Similar results are caused by overriding `buildCommand` (ex. https://github.com/NixOS/nixpkgs/pull/96552).

But it might not be an issue for tools like this (not sure, maybe you want to run at least `patchShebangsAuto` in `fixupPhase`.

DavHau replied:
fixed in 147d9a50ffe51e7fa1b853001c17d23079926ad1

jonringer reviewed (commented):
Interesting, and I like the goal of it.

Before merging (and others have had a chance to look at it), I would like for some documentation to be written. Should increase discoverability as well :).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/mach-nix-create-python-environments-quick-and-easy/6858/90

jonringer commented:
one potential issue I could see is platform specific wheels causing the hash to change. But that's already a known limitation of FOD's that are capturing more than just source.

DavHau commented:
Thanks for the feedback. I will add some docs soon.
I agree, there are potentially a few things that could change the hash. Also the resolver in pip is updated sometimes. An update of pip could result in a changed output as well.
It would be good if inputs of FODs would affect the hash as well. Wasn't there an RFC about this?

jonringer commented:
> It would be good if inputs of FODs would affect the hash as well. Wasn't there an RFC about this?

I mean like, hash was generated on `x86_64-linux`, and a mac user tried to use it, they would likely get a hash-mismatch on non-pure python wheels.

bjornfor commented on pkgs/development/interpreters/python/fetch-python-requirements.nix:
```diff
@@ -0,0 +1,127 @@
+
+          # wait for proxy to come up
+          while sleep 0.5; do
+            timeout 5 curl --proxy http://localhost:$proxyPort http://pypi.org 2>/dev/null && break
```
Add curl --fail flag? (Maybe it's not stricktly needed, but feels like a good idea to me.)

DavHau replied:
As I understand it, --fail just removes http error messages from the output. But since we have 2>/dev/null, it wouldn't make any difference at all, or would it? Also, the errors we are receiving here are `connection refused` errors which are not affected by this setting anyways.

bjornfor replied:
The `--fail` flag also makes curl exit with an error (22) instead of success (0) if there is an e.g. 404. So it's generally a good thing to use in scripts.

> Also, the errors we are receiving here are connection refused errors which are not affected by this setting anyways.

Ok. I'm just always on the alert when I see curl commands without `--fail`. I would personally add it, for peace of mind :smile:

DavHau replied:
Oh, I wasn't aware of the error codes. Fixed in 371cd6bffecb0e64cfece6438bfc13d47d63fe7f

groodt commented:
This is fantastic! Implementation and trade-offs with FOD are LGTM to me. 

I mostly have a question: is it possible to use `pip wheel` instead of `pip download` here?

DavHau committed:
fetchPythonRequirements: add targetSystem argument

FRidh commented:
I can understand the convenience, however, I do not think we should have this in Nixpkgs. The problem with this, as well as the Rust and Go examples you gave, is that the semantics to fetch the artifacts not trivial is. It won't be trivial to put items in a cache, so Nix can then fetch from them. For reproducibility this aspect is important as well. It is also the reason the Rust and Go functions should be replaced with those that have simpler fetching semantics.

There are some topics regarding this on this tracker and on Discourse.

DavHau committed:
fetchPythonRequirements: do not set phases

DavHau commented:
> I mean like, hash was generated on `x86_64-linux`, and a mac user tried to use it, they would likely get a hash-mismatch on non-pure python wheels.

Pip allows to set `--platform` which we can use to make the output independent of the buildPlatform.
This only works for wheel releases, though.
Please see bda06302cd1454488fb6c31b0ff74cefaee0f83e. I added two arguments `targetSystem` and `binaryOnly`.
`targetSystem` is intended to receive a constant string like "x86_64-linux".
It then checks that targetPlatform == buildPlatform to prevent the scenario you mentioned.
Alternatively `binaryOnly = true` can be set, in which case the targetSystem will be translated to the correct pip flags, and the fetcher result will be independent of the buildPlatform.


> This is fantastic! Implementation and trade-offs with FOD are LGTM to me.
> 
> I mostly have a question: is it possible to use `pip wheel` instead of `pip download` here?

I guess we could, but as of my understanding, the more logic we put into a FOD, the higher the risk of irreproducible results. I think the fetcher should only contain the minimum logic that get's the files downloaded. Everything else should happen in later phases.



> I can understand the convenience, however, I do not think we should have this in Nixpkgs. The problem with this, as well as the Rust and Go examples you gave, is that the semantics to fetch the artifacts not trivial is. It won't be trivial to put items in a cache, so Nix can then fetch from them. For reproducibility this aspect is important as well. It is also the reason the Rust and Go functions should be replaced with those that have simpler fetching semantics.

I can see that using that fetcher for packages in nixpkgs might not be the way to go. But if it is useful for downstream scenarios, wouldn't nixpkgs still be a good place to maintain this? Or would you be worried about it slowly creeping in over time, since people see it is there? We could make a comment in the docs: "Do not commit derivations to nixpkgs which use this function."

DavHau committed:
fetchPythonRequirements: add --fail to curl

DavHau committed:
fetchPythonRequirements: enforce targetSystem arg

SuperSandro2000 commented on pkgs/development/interpreters/python/fetch-python-requirements.nix:
```diff
@@ -0,0 +1,181 @@
+          cd $out
+          for f in $(ls $out); do
+            if [[ "$f" == *.whl ]]; then
+              pname=$(echo "$f" | cut -d "-" -f 1 | sed 's/_/-/' | sed 's/\(.*\)/\L\1/')
```
```suggestion
              pname=$(echo "$f" | cut -d "-" -f 1 | sed -e 's/_/-/' -e 's/\(.*\)/\L\1/')
```

SuperSandro2000 commented on pkgs/development/interpreters/python/fetch-python-requirements.nix:
```diff
@@ -0,0 +1,181 @@
+        outputHashAlgo = "sha256";
+        inherit outputHash;
+
+        nativeBuildInputs = with python3.pkgs; [ pythonWithMitmproxy pythonPip curl cacert ];
```
```suggestion
        nativeBuildInputs = with python3.pkgs; [ pythonWithMitmproxy pythonPip ] ++ [ curl cacert ];
```

DavHau replied:
Nothing in that list is from `python3.pkgs`, so I removed with `with` statement now.

SuperSandro2000 commented on pkgs/development/interpreters/python/fetch-python-requirements.nix:
```diff
@@ -0,0 +1,181 @@
+          "manylinux2014_aarch64"
+        ];
+      };
+      platforms = if sysToPlatforms ? "${targetSystem}" then sysToPlatforms."${targetSystem}" else throw ''
```
Does this work for non manylinux wheels? The only system I can think of right now that uses this is alpine and I am not sure if nixos static is that or still manylinux.

DavHau replied:
Oh, I haven't even known about linux wheels until you mentioned it. Added them via 5893d09cccbddc8c44848750554595c64e73082a

DavHau replied:
Wait a sec, do we even support `linux` wheels in nixpkgs? If not, it should probably not be listed here.

SuperSandro2000 reviewed (commented):
It would be great if it would support custom wheel houses and pypi servers like https://pypi.org/project/pypiserver/ to reuse pre-build wheels you build yourself.

I also left 2 small suggestions.

SuperSandro2000 commented on pkgs/development/interpreters/python/fetch-python-requirements.nix:
```diff
@@ -0,0 +1,181 @@
+
+          # wait for proxy to come up
+          while sleep 0.5; do
+            timeout 5 curl --fail --proxy http://localhost:$proxyPort http://pypi.org 2>/dev/null && break
```
```suggestion
            timeout 5 curl -fLsS --proxy http://localhost:$proxyPort http://pypi.org && break
```

Just a suggestion:

I use those args in many scripts and they are a good mixture between being silent, compatible and still showing important errors. Feel free to adapt.

-f -> --fail
-L -> follow redirects
-s -> silen
-S -> show errors in silent

DavHau replied:
Thanks for that. I tried this and it still seems to print messages like
`curl: (7) Failed to connect to localhost port 37467: Connection refused`.
Do we want this in the build log? Maybe it will confuse people to think something is wrong, even though this is expected.

DavHau replied:
Oh, I should read more carefully. Just needed to remove -S

DavHau replied:
kept only the -fs now, since -L somehow leads to a situation where SSL is required.

DavHau committed:
Update pkgs/development/interpreters/python/fetch-python-requirements.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

DavHau committed:
Update pkgs/development/interpreters/python/fetch-python-requirements.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

DavHau committed:
fetchPythonRequirements: fixes
- add `linux` platform wheels
- cleanup/refactor nativeBuildInputs
- fix curl args

DavHau commented:
Here an example on how packages could be fetched for different platforms.
This will produce the same result, no matter which system is used as a build host.
If the `onlyBinary` flag is removed, evaluation will throw an error for all foreign target systems.
```nix
let
  pkgs = import <nixpkgs> {};
  systems = {
    aarch64-linux = "sha256-o+wb3mIQCe2ZTKHAhGGZGB0BafuWyj8URK4UkKLc84w=";
    x86_64-linux = "sha256-VFnSMol5/HL/fzWIw/HgOXyIETLraw0RPnyC1+L47D4=";
    x86_64-darwin = "sha256-O0mHqiCNXx947dz36HrqEeRxRoQZAggG9LeSvAbbQng=";
  };
  sources = pkgs.lib.mapAttrs (targetSystem: outputHash:
    pkgs.python38.pkgs.fetchPythonRequirements {
      inherit targetSystem outputHash;
      requirements = [ "scipy" ];
      maxDate = 1619935837;
      onlyBinary = true;
    }
  ) systems;
in
[
  sources.aarch64-linux
  sources.x86_64-linux
  sources.x86_64-darwin
]
```

jonringer commented:
Another potential issue is that pypi has been known to remove versions of package to prevent "future installations" from receiving a bad release. Good in intention, but would break reproducibility for us as well.

DavHau commented:
> It would be great if it would support custom wheel houses and pypi servers like https://pypi.org/project/pypiserver/ to reuse pre-build wheels you build yourself.

@SuperSandro2000 I'm not sure if this can be accomplished. Pip uses the PEP 503 simple API to retrieve the list of available packages. This API does not contain information about the package release date. To be able to filter the packages by date, the proxy additionally queries the [warehouse json api](https://warehouse.pypa.io/api-reference/json.html). I assume that this API is not provided by alternative index servers.

It would be doable if your index only contained files for package names that also exist on pypi. But if it contains non public packages, then we needed some way to retrieve the release date.

But I agree that it would be nice to somehow be able to include pre-existing packages.
What we could do is:
 - provide the fetcher with a list of already existing .whl/.tar.gz/.zip package files
 - symlink these files into the downloads directory, so pip will recognize them as `already downloaded`
 - make the proxy filter the index responses to only contain exactly the one existing file for each associated package. (otherwise pip might decide to resolve to another file)
 
 Would this satisfy your use case somehow?

SuperSandro2000 commented:
It doesnot implement the warehouse api. Right now I only store pre builds on the server and i was just thinking about it we could easily implement it.

DavHau commented:
> Another potential issue is that pypi has been known to remove versions of package to prevent "future installations" from receiving a bad release. Good in intention, but would break reproducibility for us as well.

Since [PEP 592](https://www.python.org/dev/peps/pep-0592/), those files would probably be yanked instead of deleted. Still, if a package is yanked, pip will want to resolve to another version. This can only be prevented by providing frozen requirements to the fetcher.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/vscod-e-ium-users-you-can-now-use-extensions-without-explicitly-packaging-them-in-nixpkgs/12846/2

domenkozar commented:
> I can understand the convenience, however, I do not think we should have this in Nixpkgs. The problem with this, as well as the Rust and Go examples you gave, is that the semantics to fetch the artifacts not trivial is. It won't be trivial to put items in a cache, so Nix can then fetch from them. For reproducibility this aspect is important as well. It is also the reason the Rust and Go functions should be replaced with those that have simpler fetching semantics.
> 
> There are some topics regarding this on this tracker and on Discourse.

Not sure I follow, what makes it unreproducible? If it's pinned by the index date and output hash it should always get the same result and verify if that's the case.

DavHau commented:
> This is problematic if you want to use say an older nixpkgs, and cannot use
the binary cache.

I guess the only reason why an older nixpkgs would result in another output here would be the version of pip which might introduce a change in the resolver.
The fetcher could be adapted to first download a specific version of pip and then always use this pip to resolve the packages.

In this case it might probably be a good idea to rename the fetcher to something like `fetchPythonRequirementsPip_21_0_1` to make the dependency on pip's version very clear and force users to touch the code in case this version of the fetcher is not available in a specific nixpkgs.

But I have to admit, it is getting a bit more exotic now. If this is not wanted in nixpkgs, I can just push it to some other repo.

DavHau committed:
fetchPythonRequirements: improve reproducibility
- add args: pythonVersion, pipVersion, logicVersion
- always use specified pip version for fetching
- fix the version for mitmproxy

DavHau commented:
Actually there is more than just the pip version that could change.
Please have a look at 1026086b2fe3fc4b6207e9167f6c0369119456cc. I changed the UI to incorporate these variables to force the user to intentionally set them and prevent unaware changes.
The following arguments are added:
#### `pipVersion`
The fetcher will always first install this pip version in a venv and use this version for fetching


#### `pythonVersion`
Specify the python version to fetch packages for.
If, for example, the user calls the fetcher via `python3.pkgs.fetchPythonReq...` and later `python3` starts pointing to another python version, the following error would be thrown:
```
error: Attempt to use the fetchPythonRequirements package from python 3.9,
       while `pythonVersion = 3.8` was passed to it.
       Make sure to update the 'outputHash' after changing.
``` 

#### `logicVersion`
This is for future changes to the fetcher logic. Whenever we make any changes to the fetcher that could potentially change its output and are not covered by the previous arguments, we must hide the changed behind this logic switch, so that if users continue to pass `logicVersion = "1"`, they should still get the same results.

With the current design, the logic version is not exposed and instead included in the attribute name. Now the fetcher is renamed to `fetchPythonRequirements_1` where 1 stands for the logic version used.

---
If any of these arguments are unset, an error is thrown that explains usage and emphasizes that these arguments should never depend on other variables.

Of course, minor version updates for python or any change in the fetcher's system dependencies could still affect the output, but I'd assume this is unlikely enough to put us on a level similar the go or rust equivalent fetchers.

DavHau committed:
fetchPythonRequirements: replace '.' with '-' in normalized names

InLaw commented:
any update on when the pending review could be?

InLaw commented:
Review requested?

DavHau commented:
My packaging automation efforts are going into in dream2nix nowadays, therefore I'm closing this issue.
